### ISSUE TITLE
apple1 cassette interface is currently broken for cassette reads.

### ISSUE THREAD

**Issue by colinhowell** [mamedev/mame#2225]
Pinging @rb6502 because this relates to his commit ee91860.

(Don't worry, this problem will be trivial to fix; it was much harder to track down. I have added comments to the two offending lines of commit ee91860 pointing out the small fixes needed. I've rebuilt and tested the apple1 code myself with these fixes, and the cassette emulation works properly.)

Right now the apple1 cassette interface may not read properly. I say "may" because the behavior is actually unpredictable from run to run of MAME. Either the interface will always read properly (given a loaded, working cassette image and the proper procedure followed for a tape read), or it will never complete a read, simply hanging after the read command is issued no matter what cassette image is used or whether the emulated system is reset.

The problem is that when the cassette interface code was refactored into a separate device, part of the device's read I/O handler (the part actually triggered by the interface's ROM code during tape read commands) had the CPU's address, 0xc100, of the 256-byte interface ROM left in as an offset to the ROM's array, even though the device handler now accesses that array completely independently from the CPU address space. Thus it ends up reading memory from somewhere outside the ROM array.

Because of how the interface ROM's tape-read routine behaves, it may still sometimes manage to work: if the adjacent bytes actually accessed by the device handler in this out-of-bounds memory happen to be different, this will be enough for tape reads to work properly. But if these bytes are equal (as would be the case if the out-of-bounds memory is zero-filled, for instance), the tape read will hang, because the interface ROM's tape-read routine will be waiting for an input signal change that it will never be able to see.

This breakage is actually three years old. Doesn't seem like many people run this driver.  :D

To be fair, this would be a hard problem to find, and since the Apple 1's actual cassette interface has a reputation for being finicky, people probably assumed that was the cause. I only stumbled across it because a few days ago I felt like fooling around with the Apple 1 in MAME again, and even then it took quite a bit of trial and error and incorrect hypotheses before I finally realized the true cause of the problem.

**Comment by colinhowell** [comment]
@rb6502 has just fixed this bug with commit 723debd. Thanks!

(Pinging @Dagarman as well; yesterday I commented on this bug on one of his first posts on his MAME blog, and I just added an update to that comment.)

**Comment by rb6502** [comment]
No problem, and I can't believe that ever worked :)

**Comment by stilett0** [comment]
Hey @colinhowell , off-topic but while we have you here:
When we went through the massive MAME open-source relicense project (approx. May 2015 to March 2016): http://mamedev.org/?p=422
we tried a variety of means but we never were able to get in touch with you during this time to find out how you would like your code you'd contributed to be licensed (we were giving people a selection of OSI/FSF-approved licenses to choose from), so after some discussion as to how to handle cases like this,  your code (and a few others) was reverted and in your case @rb6502 graciously agreed to rewrite from scratch what you'd added. It wasn't personal! 😃 We were just playing it safe, for your and our protection. We know it's a bit silly to be that thorough, but better safe than sorry.

FYI: We're 100% sure nothing copyrightable of yours remained, but in the future any additions you'd like to contribute are intended to either match the license mentioned in the header of the specific source you're editing, or the overall MAME license. http://mamedev.org/legal.html

**Comment by colinhowell** [comment]
@stilett0 Yeah, I'm sorry about that. That was purely my fault. You did reach me; I still have the e-mails filed away (I just checked). But I was just being unresponsive, partly from being in a long-lasting depressive funk. I actually would have been fine with relicensing my contributions under BSD-3-Clause, or whatever other licensing decision you might have preferred. I hadn't realized that my lack of response proved so problematic and ended up making more work for @rb6502; had that occurred to me, I might have gotten off my ass.

I do indeed understand the importance of being thorough about licensing for a project such as this. Thanks for your efforts, and for explaining what had happened. (I kind of figured you might ask, since I had popped up again out of nowhere, but like I said, I didn't realize the consequences of my unresponsiveness.)
